Unit test for api_communicate

mushtaqak opened the pull request:
This PR adds unit tests for `api_communicate.py`

[EDUCATOR-798](https://openedx.atlassian.net/browse/EDUCATOR-798)

mushtaqak commented on video_worker/api_communicate.py:
```diff
@@ -210,8 +218,8 @@ def send_val_data(self):
             if r2.status_code > 299:
+                logger.error('VAL POST/PUT Fail: VAL')
                 ErrorObject().print_error(
-                    method=self,
```
@yro I didn't know why this is used but when passed this argument an unexpected kwarg error was being thrown by `ErrorObject().print_error`. I checked `ErrorObject().print_error` it also didn't had any impementation for `method` argument.

yro replied:
@mushtaqak It may have been a typo on my part. Good catch.

mushtaqak commented:
@muzaffaryousaf @muhammad-ammar review ?

mushtaqak commented on video_worker/tests/test_instance_config.yaml:
```diff
@@ -0,0 +1,19 @@
+---
+#---#
```
I just had copied from `default_config.yaml`

muzaffaryousaf replied:
let's remove.

mushtaqak commented on video_worker/tests/test_instance_config.yaml:
```diff
@@ -0,0 +1,19 @@
+val_secret_key: 'fake-val-secret-key'
+val_username: 'fake-val-username'
+val_password: 'fake-val-password'
+...
```
I just had copied from `default_config.yaml`

muzaffaryousaf replied:
let's remove it then.

mushtaqak commented on video_worker/tests/test_api_communicate.py:
```diff
@@ -0,0 +1,237 @@
+            VideoObject=self.Video if send_video_object else None
+        )
+
+        response = communicate_api.determine_veda_pk()
```
I have left run, run_veda, run_val and __init__ methods because they are using these methods. It would be useless to write tests for them.

muzaffaryousaf replied:
is the code being covered ?

mushtaqak replied:
I don't think those methods are covered. Now the question is _Do we need to write tests just for the sake of coverage when we are actually checking methods that actually do job?_

muzaffaryousaf replied:
So are they not doing any job ? If not then i would remove them.

mushtaqak replied:
I will write the test for one missing method so that it is being covered too.

muzaffaryousaf replied:
Thanks.

mushtaqak commented:
@yro Can you please review this?

mushtaqak commented:
@muzaffaryousaf addressed feedback

mushtaqak commented:
I will add coverage, need to check how to setup

muzaffaryousaf commented:
@muhammad-ammar can help you out with coverage locally, it should't be hard to setup.

yro commented on video_worker/api_communicate.py:
```diff
@@ -104,6 +109,7 @@ def determine_veda_pk(self):
+            logger.error('VEDA API Fail: Check VEDA API config')
             ErrorObject().print_error(
                 message='VEDA API Fail: Check VEDA API config'
             )
```
@mushtaqak These ErrorObject things can be eliminated. The logger does that function better.

muzaffaryousaf replied:
I'll write another ticket to clean this up. Thanks @yro for pointing this out.

mushtaqak replied:
@muzaffaryousaf I have removed `ErrorObject` in for api_communicate.py and coverage has increased because we were not checking `ErrorObject` previously.

yro commented on video_worker/tests/test_api_communicate.py:
```diff
@@ -0,0 +1,252 @@
+    @patch('video_worker.api_communicate.requests.get')
+    @patch('video_worker.api_communicate.logger')
+    def test_send_val_data(self, data, mock_logger, mock_get, mock_post, mock_put):
+        """
```
@mushtaqak This is really nice.

muzaffaryousaf commented:
@mushtaqak there  are still few comments that needs your attention. Let me know when you're done. Thanks.

mushtaqak commented:
`nosetests --with-coverage` shows `api_communicate.py` has **81%** coverage:

```
video_worker/__init__/__init__.py                                     157    102    35%
video_worker/__init__.py                                              157    102    35%
video_worker/abstractions.py                                          127     85    33%
video_worker/api_communicate.py                                       113     21    81%
video_worker/celeryapp.py                                              37     16    57%
video_worker/config.py                                                 49     27    45%
video_worker/generate_apitoken.py                                      32     20    38%
video_worker/generate_delivery.py                                      86     64    26%
video_worker/generate_encode.py                                       152    133    13%
video_worker/global_vars.py                                            16      0   100%
video_worker/reporting.py                                              77     55    29%
video_worker/validate.py                                               84     70    17%
video_worker/video_images.py                                           81     25    69%
---------------------------------------------------------------------------------------
TOTAL                                                               40297  31191    23%
----------------------------------------------------------------------
Ran 42 tests in 1.106s

FAILED (SKIP=11, errors=1)
```

mushtaqak commented on video_worker/tests/test_api_communicate.py:
```diff
@@ -0,0 +1,242 @@
+
+        if communicate_api.VideoObject:
+            communicate_api.VideoObject.val_id = kwargs.get('val_id', None)
+            communicate_api.VideoObject.course_url = kwargs.get('course_url', [])
```
Yes it's actually a list of course ids. Used in api_comunicate.py https://github.com/edx/edx-video-worker/blob/master/video_worker/api_communicate.py#L178

muzaffaryousaf replied:
Is it the same one which is now a dict of list ?

mushtaqak replied:
hmm, according to my knowledge and understanding, I don't think so but @yro can answer better.

yro replied:
@mushtaqak Yes, this corresponds to the course url field in VAL. It's the one that recently changed.

mushtaqak commented:
After removing `ErrorObject` statements and addressing feedback, coverage report now shows **88%** for `api_communicate.py`:

```
video_worker/__init__/__init__.py                                     157    102    35%
video_worker/__init__.py                                              157    102    35%
video_worker/abstractions.py                                          127     82    35%
video_worker/api_communicate.py                                       106     13    88%
video_worker/celeryapp.py                                              37     15    59%
video_worker/config.py                                                 49     27    45%
video_worker/generate_apitoken.py                                      32     15    53%
video_worker/generate_delivery.py                                      86     63    27%
video_worker/generate_encode.py                                       152    133    13%
video_worker/global_vars.py                                            16      0   100%
video_worker/reporting.py                                              77     55    29%
video_worker/validate.py                                               84     69    18%
video_worker/video_images.py                                           81     24    70%
---------------------------------------------------------------------------------------
TOTAL                                                               40626  30818    24%
----------------------------------------------------------------------
Ran 42 tests in 1.594s

FAILED (SKIP=11, errors=9)
```

mushtaqak commented on video_worker/api_communicate.py:
```diff
@@ -258,14 +249,5 @@ def send_val_data(self):
                 return None
-
-
-def main():
```
@yro @muzaffaryousaf I don't think we are doing anything here, so better to remove this, right?

yro replied:
@mushtaqak Absolutely, remove it. Thanks!

muzaffaryousaf commented on video_worker/api_communicate.py:
```diff
@@ -52,9 +54,7 @@ def run_veda(self):
         if self.veda_token is None:
-            ErrorObject().print_error(
-                message='VEDA API Conn Fail:\nInvalid Setup/Method'
-            )
```
@yro Are we using these print statements anywhere for logs purpose OR is it safe to remove them ?

yro replied:
We're using them for logs, but, actually raising them as errors is maybe a better choice. If you want to just log the error for now, that's totally fine.

muzaffaryousaf commented on video_worker/tests/test_api_communicate.py:
```diff
@@ -0,0 +1,252 @@
+
+        if communicate_api.veda_token:
+            communicate_api.veda_headers = {
+                'Authorization': 'Token ' + communicate_api.veda_token,
```
can we use .format instead of `+`

muzaffaryousaf commented:
@mushtaqak Thanks for adding coverage.

muzaffaryousaf commented:
> After removing `ErrorObject` statements and addressing feedback, coverage report now shows **88%** for `api_communicate.py`:
```
video_worker/__init__/__init__.py                                     157    102    35%
video_worker/__init__.py                                              157    102    35%
video_worker/abstractions.py                                          127     82    35%
video_worker/api_communicate.py                                       106     13    88%
video_worker/celeryapp.py                                              37     15    59%
video_worker/config.py                                                 49     27    45%
video_worker/generate_apitoken.py                                      32     15    53%
video_worker/generate_delivery.py                                      86     63    27%
video_worker/generate_encode.py                                       152    133    13%
video_worker/global_vars.py                                            16      0   100%
video_worker/reporting.py                                              77     55    29%
video_worker/validate.py                                               84     69    18%
video_worker/video_images.py                                           81     24    70%
---------------------------------------------------------------------------------------
TOTAL                                                               40626  30818    24%
----------------------------------------------------------------------
Ran 42 tests in 1.594s
FAILED (SKIP=11, errors=9)
```


Perfect 💯

yro commented:
@mushtaqak @muzaffaryousaf I never thought I'd see the day this repo had such good coverage. Really nice work, thank you.
